TSGuild-99: Create a TypeScript version of election-api-javascript

rossanodan opened the pull request:
## What?

This pull request adds a new project to the list, `election-api-typescript`. This is an exact copy of `election-api-javascript` with the only difference that the development environment is based on TypeScript.

## Why?

This will help recruitment to test candidates' TypeScript capabilities and it will also give more choice to the candidates who prefer writing TypeScript code.

## Next steps
Create `election-ui-typescript`

@bbc/typescript-guild

Commit:
TSGuild-99 set up TypeScript project for election-api-javascript

This commit creates the skeleton application of election-api-javascript in a TypeScript environment.

Commit:
TSGuild-99 set up Jest configuration for TypeScript and copy the test files

Added Jest configuration for TypeScript support and copied JSON test files from election-api-javascript.

Commit:
TSGuild-99 Copy README and tasks with interview instructions

rossanodan commented on election-api-typescript/src/resultsController.ts:
```diff
@@ -0,0 +1,24 @@
+  return store.getResult(id);
+};
+
+// TODO: Ask someone why this returns empty object instead of result and add type
+const newResult = (result: Result) => {
+  store.newResult(result);
+  return {};
+};
```
It is not clear why this returns an empty object instead of the `result`. Any ideas?

paulbrimicombe replied:
I don't think there is a good reason to be honest. There are lots of things you _could_ return that would be sensible. I'd probably either return nothing (i.e. `null`) and rely on the status code indicating that the write was successful or something with a message saying that the result was created successfully.

amckinlaybbc commented on election-api-typescript/package.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "hiring-api-test",
+  "version": "1.0.0",
+  "description": "Node.jsn and TypeScript coding test",
```
```suggestion
  "description": "Node.js and TypeScript coding test",
```

amckinlaybbc commented on election-api-typescript/README.md:
````diff
@@ -0,0 +1,54 @@
+- npm - will be installed with node, or see https://docs.npmjs.com/downloading-and-installing-node-js-and-npm
+
+```
+npm install
+npm start
````
As it stands wouldn't this need to be
```suggestion
npm install
npm build
npm start
```
?

What I think might be better is to install and use [`ts-node-dev`](https://www.npmjs.com/package/ts-node-dev) and replace the `start` command in the `package.json` file with:

`ts-node-dev --respawn --transpile-only ./src/index.ts`

Which removes the need for the build step and keeps the setup instructions the same between the JS version and this version.

paulbrimicombe commented on election-api-typescript/tsconfig.json:
```diff
@@ -0,0 +1,22 @@
+    "forceConsistentCasingInFileNames": true,
+
+    /* Type Checking */
+    "strict": true,
```
I think it would be good to add `noImplicitAny` here as well

paulbrimicombe commented on election-api-typescript/src/resultsController.ts:
```diff
@@ -0,0 +1,24 @@
+
+const store = resultStore();
+
+const getResult = (id: number): Result | undefined => {
```
I know that this is basically a port of the JS version but I'd much prefer to use `null` to indicate missing items than `undefined` if possible.

```suggestion
const getResult = (id: number): Result | null => {
```

paulbrimicombe commented on election-api-typescript/src/types.ts:
People I work with know that I have a dislike for `types` dumping ground files. I think it would be much better to place the types inside the files where they originate (in this case inside the results service). This is definitely a matter of taste though!

paulbrimicombe reviewed (approved):
Looks good to me. I've added a few comments. Thanks for taking the time to do this.

Commit:
Update election-api-typescript/package.json

Co-authored-by: Allan Mckinlay <122284042+amckinlaybbc@users.noreply.github.com>

Commit:
Update election-api-typescript/package.json

Co-authored-by: Paul Douglas Brimicombe <paul.brimicombe@bbc.co.uk>

Commit:
Update election-api-typescript/package.json

Co-authored-by: Paul Douglas Brimicombe <paul.brimicombe@bbc.co.uk>

Commit:
Update election-api-typescript/src/resultsController.ts

Co-authored-by: Paul Douglas Brimicombe <paul.brimicombe@bbc.co.uk>

Commit:
TSGuild-99 return null and a message to terminal when adding data to the store

As per comment, I'm returning null to rely on a message that indicates the new data has been created successfully.

Commit:
TSGuild-99 fix return type function getResult

Result | null was incompatible with the result type of the function.

Commit:
Update election-api-typescript/jest.config.ts

Co-authored-by: Paul Douglas Brimicombe <paul.brimicombe@bbc.co.uk>

Commit:
TSGuild-99 installed and used ts-node-dev to startup the local server

Commit:
TSGuild-99 add noImplicitAny to tsconfig rules

paulbrimicombe reviewed (approved):
👍
